### ISSUE TITLE
APC_AOS Status Failed

### ISSUE THREAD

**Issue by fatpacket** [ytti/oxidized#2179]
have added an APC NMS card to the router.db and done the mapping in the config file.  I have created a local user on the APC card and enabled the FTP server on the card.  I can successfully connect and pull the config.ini from the APC card manually with the configured account but when I try to run it through Oxidized the process fails. It does not generate a crash. It seems to create an entry in the local repo but never adds the config.ini to the configs there.  See Below:

D, [2020-09-08T18:01:22.606355 #13293] DEBUG -- : lib/oxidized/worker.rb: Added /apc-nms to the job queue
D, [2020-09-08T18:01:22.606721 #13293] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-09-08T18:01:22.607172 #13293] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for apc-nms at 2020-09-08 22:01:22 UTC
D, [2020-09-08T18:01:22.607601 #13293] DEBUG -- : lib/oxidized/job.rb: Config fetched for apc-nms at 2020-09-08 22:01:22 UTC
W, [2020-09-08T18:01:23.607523 #13293]  WARN -- : /apc-nms status fail, retry attempt 1
D, [2020-09-08T18:01:23.607975 #13293] DEBUG -- : lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 8 of 9
D, [2020-09-08T18:01:23.608335 #13293] DEBUG -- : lib/oxidized/worker.rb: Added /apc-nms to the job queue
D, [2020-09-08T18:01:23.608563 #13293] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-09-08T18:01:23.609038 #13293] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for apc-nms at 2020-09-08 22:01:23 UTC
D, [2020-09-08T18:01:23.609447 #13293] DEBUG -- : lib/oxidized/job.rb: Config fetched for apc-nms at 2020-09-08 22:01:23 UTC
W, [2020-09-08T18:01:24.609242 #13293]  WARN -- : /apc-nms status fail, retry attempt 2
D, [2020-09-08T18:01:24.609616 #13293] DEBUG -- : lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 8 of 9
D, [2020-09-08T18:01:24.609987 #13293] DEBUG -- : lib/oxidized/worker.rb: Added /apc-nms to the job queue
D, [2020-09-08T18:01:24.610329 #13293] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-09-08T18:01:24.610792 #13293] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for apc-nms at 2020-09-08 22:01:24 UTC
D, [2020-09-08T18:01:24.611203 #13293] DEBUG -- : lib/oxidized/job.rb: Config fetched for apc-nms at 2020-09-08 22:01:24 UTC
W, [2020-09-08T18:01:25.610952 #13293]  WARN -- : /apc-nms status fail, retry attempt 3
D, [2020-09-08T18:01:25.611276 #13293] DEBUG -- : lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 8 of 9
D, [2020-09-08T18:01:25.611632 #13293] DEBUG -- : lib/oxidized/worker.rb: Added /apc-nms to the job queue
D, [2020-09-08T18:01:25.611965 #13293] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2020-09-08T18:01:25.612378 #13293] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for apc-nms at 2020-09-08 22:01:25 UTC
D, [2020-09-08T18:01:25.612782 #13293] DEBUG -- : lib/oxidized/job.rb: Config fetched for apc-nms at 2020-09-08 22:01:25 UTC
W, [2020-09-08T18:01:26.612684 #13293]  WARN -- : /apc-nms status fail, retries exhausted, giving up
D, [2020-09-08T18:01:26.613064 #13293] DEBUG -- : lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 9 of 9
D, [2020-09-08T18:01:26.613365 #13293] DEBUG -- : lib/oxidized/worker.rb: Running :nodes_done hook

router.db:
apc-nms:192.168.27.7:apc

config file:
---
username: ******
password: ******
model: cisco
resolve_dns: true
interval: 300
use_syslog: false
debug: true
threads: 30
timeout: 50
retries: 3
prompt: !ruby/regexp /^([\w.@-]+[#>]\s?)$/
rest: 127.0.0.1:8888
next_adds_job: false
vars: {}
groups: {}
models: {}
pid: "/home/oxidized/.config/oxidized/pid"
crash:
  directory: "/home/oxidized/.config/oxidized/crashes"
  hostnames: false
stats:
  history_size: 10
input:
  default: ssh
  debug: false
  ssh:
    secure: false
  ftp:
    passive: true
  utf8_encoded: true
output:
  default: git
  file:
    directory: "/home/oxidized/configs"
  git:
    user: Oxidized
    email: oxidized@local
    repo: "/home/oxidized/configs/device-config-repo.git"
source:
  default: csv
  csv:
    file: "/home/oxidized/.config/oxidized/router.db"
    delimiter: !ruby/regexp /:/
    map:
      name: 0
      ip: 1
      model: 2
    vars_map:
      enable: 3
    gpg: false
model_map:
  cisco: ios
  nexus: nxos
  asa: asa
  wireless: aireos
  apc: apc_aos
hooks:
  push_to_remote:
    type: githubrepo
    events: [post_store]
    remote_repo: "git@github.com/*******/device-config-repo.git"
    publickey: "/home/oxidized/.ssh/id_rsa.pub"
    privatekey: "/home/oxidized/.ssh/id_rsa"

**Comment by flea100** [comment]
We have the same issue.  It would be good to get some traction on a fix.

**Comment by NuNetAdmins** [comment]
We have the same issue as well.  The APC component is useless at this time.  I would be happy to perform any testing if needed to help with a fix for this problem.

**Comment by mortzu** [comment]
Can you please reformat your paste to make it readable? (https://docs.github.com/en/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code)

**Comment by FostWare** [comment]
I'm not sure whether this helps, but I think the Dockerfile is missing gem packages for `net-ftp` and  `io-wait` (it asks for version 0.1.0).
I then get an error with "550 RETR requested action not taken.", but I think that's another issue to diagnose.

**Comment by FostWare** [comment]
Confirmed adding net-ftp and io-wait to the Docker file works as expected.

`gem install io-wait:0.1.0 net-ftp` would occasionally still flag io-wait version incompatibility, so I added a separate section to Dockerfile to:
1) ensure io-wait is installed first
2) clarify that io-wait:0.1.0 is dependent on Ruby 2.5.0

The "550 RETR..." error listed above, was related to passive mode requirement on both APAP7921 v3.9.2 and AP8858 v6.8.2 NMCs
```
groups:
...
  apc_aos:
    username: ********
    password: ********
input:
...
  ftp:
    passive: true
model_map:
...
  apc: apc_aos
```

**Comment by theturtle626** [comment]
I have the same error , I install oxidized on ubuntu (Raspberry Pi 4), I can get config from RouterOS, then I add another APC NMS card, connect by ftp service with an account/password without problem, I also did the following install and config

- gem install io-wait:0.1.0 net-ftp

- Modify in config file
groups:
...
  apc_aos:
    username: ********
    password: ********
input:
...
  ftp:
    passive: true
model_map:
...
  apc: apc_aos

- Log file error
WARN -- : /smc1000tw status fail, retry attempt 1
D, [2023-01-04T16:56:08.155513 #3371586] DEBUG -- : lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 1 of 2
D, [2023-01-04T16:56:08.155784 #3371586] DEBUG -- : lib/oxidized/worker.rb: Added /smc1000tw to the job queue
D, [2023-01-04T16:56:08.156230 #3371586] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2023-01-04T16:56:08.156048 #3371586] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for smc1000tw at 2023-01-04 16:56:08 +0800
D, [2023-01-04T16:56:08.156607 #3371586] DEBUG -- : lib/oxidized/job.rb: Config fetched for smc1000tw at 2023-01-04 16:56:08 +0800
W, [2023-01-04T16:56:09.158035 #3371586]  WARN -- : /smc1000tw status fail, retry attempt 2
D, [2023-01-04T16:56:09.158272 #3371586] DEBUG -- : lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 1 of 2
D, [2023-01-04T16:56:09.158512 #3371586] DEBUG -- : lib/oxidized/worker.rb: Added /smc1000tw to the job queue
D, [2023-01-04T16:56:09.158908 #3371586] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2023-01-04T16:56:09.158737 #3371586] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for smc1000tw at 2023-01-04 16:56:09 +0800
D, [2023-01-04T16:56:09.159336 #3371586] DEBUG -- : lib/oxidized/job.rb: Config fetched for smc1000tw at 2023-01-04 16:56:09 +0800
W, [2023-01-04T16:56:10.160793 #3371586]  WARN -- : /smc1000tw status fail, retry attempt 3
D, [2023-01-04T16:56:10.161038 #3371586] DEBUG -- : lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 1 of 2
D, [2023-01-04T16:56:10.161323 #3371586] DEBUG -- : lib/oxidized/worker.rb: Added /smc1000tw to the job queue
D, [2023-01-04T16:56:10.161731 #3371586] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2023-01-04T16:56:10.161559 #3371586] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for smc1000tw at 2023-01-04 16:56:10 +0800
D, [2023-01-04T16:56:10.162101 #3371586] DEBUG -- : lib/oxidized/job.rb: Config fetched for smc1000tw at 2023-01-04 16:56:10 +0800
W, [2023-01-04T16:56:11.163602 #3371586]  WARN -- : /smc1000tw status fail, retries exhausted, giving up
